pkgs/development/python-modules/mmcv/default.nix

```diff
@@ -0,0 +1,92 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, pytest
+, pythonOlder
+, cudaSupport ? false
+, cudatoolkit
```

Comment:
Not that I'd consider this a blocker for this PR, but for future work consider using `cudaPackages` over `cudatoolkit`. `cudaPackages` uses redistributables from NVIDIA and allows for much smaller closures because you choose which libraries you need. It also makes it easier to maintain because we then have greater visibility into what the package is using.

For an example, take a look at https://github.com/NixOS/nixpkgs/blob/7f194c741b122171494cb8852020310fa57a2944/pkgs/development/python-modules/torchvision/default.nix#L17-L40.

I break the dependencies into those needed only at build time, run time, or both (common). Unfortunately, because most packages expect all the CUDA libraries to be in a single directly, we need to use `symlinkJoin` to group them together.

Also, make sure you're specifying the right C/C++ compilers if you're building with CUDA support, like here: https://github.com/NixOS/nixpkgs/pull/218265. (Torchvision doesn't yet do that; I've got an open issue here: https://github.com/NixOS/nixpkgs/issues/221898.)

The compiler provided by `stdenv` isn't necessarily supported by NVCC, and it's possible to get odd symbol errors during linking if the other derivations were built with a different compiler or language standard. That's why it's important to use `backendStdenv.cc`.

Torchvision and Torch will probably have other things you might want to look at. If you have any questions, don't hesitate to reach out to me or the other CUDA maintainers on Matrix: https://matrix.to/#/#cuda:nixos.org.

Reply:
hey @ConnorBaker , thanks a lot for the advices and examples. thanks @bcdarwin  for raising this. I am thinking about using torch.cudaCapabilities as well. but I think there is a valid usecase for people to overwrite with torch-bin (doesn't need to compile the torchWithCuda), which doesn't have passthru for cudaCapabilities. would you suggest that I copy the same pattern from torch/default.nix to get cudaCababilities?

as for the error, I am already getting it with opencv4, as it is compiled with gcc 12, and cudatoolkit uses gcc 11. what is the best to deal with needing to compile with cudatoolkit.cc only for the build phase. use stdenv.cc for the derivation.

Reply:
also. with following nix-shell, the torchvision fails to build with cuda.

```
with import ../nixpkgs {
  config.allowUnfree = true;
 };
let
  mypthon =
    let
      packageOverrides = self: super: {
        # packagesOverride, will affect all packages that uses the resulted packages as deps
        torch = super.torch.override{ magma = magma-cuda; cudaSupport = true; };
        # opencv4 = super.opencv4.override {
        #   enableBlas = false;
        # };
      };
    in
    python3.override { inherit packageOverrides; self = mypthon; };
in
let
  pythonEnv = mypthon.withPackages (ps: with ps; [ torchvision ]);
in
mkShell {
  packages = [ pythonEnv  ];
}
    


 raise RuntimeError(
RuntimeError: The current installed version of g++ (12.2.0) is greater than the maximum required version by CUDA 11.7 (11.5.0). Please make sure to use an adequate version of g++ (>=6.0.0, <=11.5.0).
/nix/store/c3f4jdwzn8fm9lp72m91ffw524bakp6v-stdenv-linux/setup: line 1593: pop_var_context: head of shell_variables not a function context
        
```

Reply:
> but I think there is a valid usecase for people who want to overwrite with torch-bin (doesn't require compilation), but doesn't have passthru for cudaCapabilities.

IIRC, newer torch binaries use CUDA redistributables from PyPi. If they haven't moved over to that yet, then their builder is still using bash scripts to manually copy libraries into their binary :(

There's a hardcoded set of supported capabilities for each version of PyTorch. You can find them here: https://github.com/pytorch/pytorch/blob/49444c3e546bf240bed24a101e747422d1f8a0ee/torch/utils/cpp_extension.py#L1751-L1752.

Because it's a binary distributions and not "built" with Nixpkgs, the CUDA capabilities supported by the binary aren't tied to what we specify in our config.

For that reason I'd recommend you use the same logic `torch/default.nix` has. (Those versions are actually sourced from that same file!)

> as for the odd symbol error, I am already getting that with opencv4, as it is compiled with stdenv.cc gcc 12, and cudatoolkit uses gcc 11. I will give it another go.

I have a PR open which should help with that: https://github.com/NixOS/nixpkgs/pull/221370.

> i assume gcc 12.2 is from stdenv

That's right, it is :(

I know that error well. It's from here: https://github.com/pytorch/pytorch/blob/49444c3e546bf240bed24a101e747422d1f8a0ee/torch/utils/cpp_extension.py#L415. (Related discussion on whether checks like that should be removed when included in Nixpkgs: https://github.com/NixOS/nixpkgs/issues/221564#issuecomment-1478486895.)

That's odd that you're getting it though, since the `torchvision` derivation should correctly set the CC/CXX environment variables to the right compiler: https://github.com/ConnorBaker/nixpkgs/blob/f1d538b45955758f7e1b2e73ab4ee3857a710c9a/pkgs/development/python-modules/torchvision/default.nix#L67-L72.

With respect to your nix file can you try doing something like the following instead? Substitute whatever capability is appropriate for your device (though using just one does make it a lot faster to rebuild stuff from source!).

```nix
with import ../nixpkgs {
  config = {
    allowUnfree = true;
    cudaSupport = true;
    cudaCapabilities = [ "8.6" ];
  };
};
```

I think it was choosing the wrong compiler because `cudaSupport` wasn't explicitly set to `true`.

Let me know how that works!

Reply:
@ConnorBaker, your MR on opencv fixed my symbol error.

Reply:
Yay! Now it just needs to get merged (somewhat difficult given the large number of rebuilds)!